pkg/healthcheck/healthcheck.go

```diff
@@ -352,6 +353,33 @@ func (hc *HealthChecker) allCategories() []category {
 						return hc.checkNamespace(ctx, hc.ControlPlaneNamespace, true)
 					},
 				},
+				{
+					description: "correct pod security policies",
+					hintAnchor:  "l5d-existence-psp",
+					fatal:       true,
+					check: func(ctx context.Context) error {
+						var err error
+						var controlPlaneReplicaSet []v1beta1.ReplicaSet
+						controlPlaneReplicaSet, err = hc.kubeAPI.GetReplicaSets(ctx, hc.httpClient, hc.ControlPlaneNamespace)
```

Comment:
Any reasons why we want to use replica sets, not just pods?

Reply:
With replica set, it will be useful to compare (desired) [`Replicas`](https://godoc.org/k8s.io/api/apps/v1#ReplicaSetSpec) with [`ReadyReplicas`](https://godoc.org/k8s.io/api/apps/v1#ReplicaSetStatus).

Reply:
@ihcsim PSP can forbid a pod to be created, so we can only check the failure in the Replicaset.

Reply:
> PSP can forbid a pod to be created

That makes sense, since the PSP is an admission controller.

In that case, can we check for the `v1beta1.ReplicaSetReplicaFailure` error instead? Looking at the [doc](https://godoc.org/k8s.io/api/apps/v1#ReplicaSetConditionType), the replica set condition will have this `Type` if the _pods fails to be created due to insufficient quota, limit ranges, pod security policy, node selectors, etc. or deleted due to kubelet being down or finalizers are failing_.

```diff
diff --git a/pkg/healthcheck/healthcheck.go b/pkg/healthcheck/healthcheck.go
index e7f4247f..a24c118c 100644
--- a/pkg/healthcheck/healthcheck.go
+++ b/pkg/healthcheck/healthcheck.go
@@ -976,19 +976,18 @@ func checkUnschedulablePods(pods []corev1.Pod) error {
 	return nil
 }
 
-func checkPodSecurityPolicies(rst []v1beta1.ReplicaSet) error {
-	var replicaSets []string
+func checkControlPlaneReplicaSets(rst []v1beta1.ReplicaSet) error {
+	var errors string
 	for _, rs := range rst {
 		for _, r := range rs.Status.Conditions {
-			if strings.Contains(r.Message, "unable to validate against any pod security policy") {
-				replicaSets = append(replicaSets, rs.Name)
+			if r.Type == v1beta1.ReplicaSetReplicaFailure && r.Status == v1beta1.ConditionTrue {
+				errors += fmt.Sprintf("%s: %s, %s\n", rs.Name, r.Reason, r.Message)
 			}
 		}
 	}
 
-	if len(replicaSets) != 0 {
-		replicasets := strings.Join(replicaSets, ",")
-		return fmt.Errorf("Invalid pod security policy for the replicasets %s", replicasets)
+	if errors != "" {
+		return fmt.Errorf("errors in creating control plane replicas: ", errors)
 	}
 
 	return nil
```

Reply:
> ```diff
> v1beta1.ConditionTrue
> ```
i couldn't find this in the [doc](https://godoc.org/k8s.io/api/apps/v1beta1)

Reply:
Sorry, it's `corev1.ConditionTrue`.

Reply:
i thiink it should be `corev1.ConditionTrue `

Reply:
The output of the above code is something like this

```
× correct pod security policies
    errors in creating control plane replicas linkerd-controller-8647d7d878: FailedCreate, pods "linkerd-controller-8647d7d
878-" is forbidden: unable to validate against any pod security policy: []
linkerd-controller-cd8f8d4c5: FailedCreate, pods "linkerd-controller-cd8f8d4c5-" is forbidden: unable to validate against a
ny pod security policy: []
linkerd-grafana-7f8b57bd95: FailedCreate, pods "linkerd-grafana-7f8b57bd95-" is forbidden: unable to validate against any p
od security policy: []
linkerd-identity-67cfbdf4d5: FailedCreate, pods "linkerd-identity-67cfbdf4d5-" is forbidden: unable to validate against any pod security policy: []
linkerd-prometheus-64b89cd5cd: FailedCreate, pods "linkerd-prometheus-64b89cd5cd-" is forbidden: unable to validate against any pod security policy: []
linkerd-sp-validator-8465d9789d: FailedCreate, pods "linkerd-sp-validator-8465d9789d-" is forbidden: unable to validate against any pod security policy: []
linkerd-web-6cd6965f77: FailedCreate, pods "linkerd-web-6cd6965f77-" is forbidden: unable to validate against any pod security policy: []
 
    see https://linkerd.io/checks/#l5d-existence-psp for hints

Status check results are ×
```

I dont think that we need to show Reason and Message for each pods as we know it is same for all the pods i.e. `Invalid pod security policy`

Reply:
Yup, fine by me.